File: scripts/github-actions/cleanup_old_data.py
#!/usr/bin/env python3
"""古いデータを定期削除するスクリプト（30日保持）"""

import os
import sys
from datetime import datetime, timedelta
import psycopg2

# データ保持期間
RETENTION_DAYS = 30


def get_database_url() -> str:
    url = os.environ.get("DATABASE_URL")
    if not url:
        print("Error: DATABASE_URL environment variable not set")
        sys.exit(1)
    return url


def cleanup_old_data():
    conn = psycopg2.connect(get_database_url())
    try:
        cutoff_date = datetime.utcnow() - timedelta(days=RETENTION_DAYS)
        print(f"Cleanup started at {datetime.now().isoformat()}")
        print(f"Retention: {RETENTION_DAYS} days")
        print(f"Cutoff date: {cutoff_date.date()}")
        print("-" * 50)

        total_deleted = 0

        with conn.cursor() as cur:
            # 1. StockAnalysis（ポートフォリオ分析）
            cur.execute('SELECT COUNT(*) FROM "StockAnalysis" WHERE "analyzedAt" < %s', (cutoff_date,))
            count = cur.fetchone()[0]
            print(f"\n[1/3] StockAnalysis: {count} records to delete")
            if count > 0:
                cur.execute('DELETE FROM "StockAnalysis" WHERE "analyzedAt" < %s', (cutoff_date,))
                print(f"  Deleted: {cur.rowcount}")
                total_deleted += cur.rowcount

            # 2. PurchaseRecommendation（ウォッチリスト購入推奨）
            cur.execute('SELECT COUNT(*) FROM "PurchaseRecommendation" WHERE date < %s', (cutoff_date.date(),))
            count = cur.fetchone()[0]
            print(f"\n[2/3] PurchaseRecommendation: {count} records to delete")
            if count > 0:
                cur.execute('DELETE FROM "PurchaseRecommendation" WHERE date < %s', (cutoff_date.date(),))
                print(f"  Deleted: {cur.rowcount}")
                total_deleted += cur.rowcount

            # 3. UserDailyRecommendation（あなたへのおすすめ）
            cur.execute('SELECT COUNT(*) FROM "UserDailyRecommendation" WHERE date < %s', (cutoff_date.date(),))
            count = cur.fetchone()[0]
            print(f"\n[3/3] UserDailyRecommendation: {count} records to delete")
            if count > 0:
                cur.execute('DELETE FROM "UserDailyRecommendation" WHERE date < %s', (cutoff_date.date(),))
                print(f"  Deleted: {cur.rowcount}")
                total_deleted += cur.rowcount

        conn.commit()
        print("\n" + "=" * 50)
        print(f"Cleanup completed! Total deleted: {total_deleted} records")

    except Exception as e:
        print(f"Error during cleanup: {e}")
        sys.exit(1)
    finally:
        conn.close()


if __name__ == "__main__":
    cleanup_old_data()

File: scripts/github-actions/cleanup-old-data.ts
#!/usr/bin/env npx tsx
/**
 * 古いデータを定期削除するスクリプト
 *
 * 削除対象:
 * - StockAnalysis: 1週間より古いデータ
 *
 * 週1回（日曜日）に実行を想定
 */

import { PrismaClient } from "@prisma/client"

const prisma = new PrismaClient()

async function cleanupOldData(): Promise<void> {
  try {
    // 削除基準日（1週間前）
    const oneWeekAgo = new Date()
    oneWeekAgo.setDate(oneWeekAgo.getDate() - 7)

    console.log(`Cleanup started at ${new Date().toISOString()}`)
    console.log(`1 week ago: ${oneWeekAgo.toISOString().split("T")[0]}`)
    console.log("-".repeat(50))

    let totalDeleted = 0

    // StockAnalysis（1週間より古いデータ）
    console.log("\n[1/1] Cleaning up StockAnalysis...")

    const countBefore = await prisma.stockAnalysis.count({
      where: {
        analyzedAt: {
          lt: oneWeekAgo,
        },
      },
    })
    console.log(`  Records to delete: ${countBefore}`)

    if (countBefore > 0) {
      const result = await prisma.stockAnalysis.deleteMany({
        where: {
          analyzedAt: {
            lt: oneWeekAgo,
          },
        },
      })
      totalDeleted += result.count
      console.log(`  Deleted: ${result.count} records`)
    }

    console.log("\n" + "=".repeat(50))
    console.log(`Total deleted: ${totalDeleted} records`)
    console.log("Cleanup completed successfully!")
  } catch (error) {
    console.error(`Error during cleanup: ${error}`)
    process.exit(1)
  } finally {
    await prisma.$disconnect()
  }
}

cleanupOldData()

export {}
